fix: no longer crash on extensions-ception

ndrsg opened the pull request:
fix handling of extensions-tags inside extensions-tags by counting the depth

- [x] I agree to follow the project's [code of conduct](https://github.com/georust/geo/blob/master/CODE_OF_CONDUCT.md).
- [ ] I added an entry to `CHANGELOG.md` if knowledge of this change could be valuable to users.
---

lnicola commented on src/parser/extensions.rs:
```diff
@@ -57,4 +56,23 @@ mod tests {
+            </extensions>",
+            GpxVersion::Gpx11
+        );
+        dbg!(&result);
```
```suggestion
```

lnicola commented:
Can you please squash if it's not too much of a hassle?

ndrsg commented:
sure :)

urschrei commented:
Just want to note that I very much enjoyed "extensions-ception"

lnicola commented:
I can't check right now, but can there be two closing tags for only one opening tag? What happens in that case?

ndrsg commented:
1 opening followed by 2 closing tags:

This extensions-consume function would exit on the first closing tag without error.

This would lead to the outer logic to receive a extensions-closing tag. Depends if this case is handled there.

Might happen with other non-gpx-tags inside an extensions-block as well.

lnicola commented:
But there's no code to return out of the function when the depth reaches 0. The old version used to do that.

lnicola commented on src/parser/extensions.rs:
```diff
@@ -11,32 +11,31 @@ use crate::parser::Context;
             XmlEvent::EndElement { name, .. } => {
                 if &name.local_name == "extensions" {
-                    return Ok(());
+                  depth -= 1;
```
Might want to return here if the depth reached 0.

ndrsg commented:
I will have a look tomorrow.

Another option might be recursion instead of counting depth?

ndrsg commented:
soo.. I changed the approach a bit and oriented on the other implementations, also keeping in mind a future implementation of parsing the extensions content into a structure with kind of "generic parser"?

also added some tests, where I would expect errors. turned out, that the xmlreader already tests the xml-structure quite well.

now treating "extensions"-tags inside an "extensions"-block just like any other tag, hence its just coincidence, if an internal tag is called "extensions", right?

lnicola commented on src/parser/extensions.rs:
```diff
@@ -9,40 +9,53 @@ use xml::reader::XmlEvent;
+            XmlEvent::EndElement { name } => {
+                // as long as there is an inner tag open..
+                if let Some(current_inner_tag) = inner_tag_stack.pop() {
+                    // the closing tag has to match the current open tag, this makes it also impossible to close the "extensions" while an inner tag is open
```
If the XML library doesn't allow a mismatched closing tag, I think we shouldn't handle it here. Then the stack isn't necessary.

Instead I was thinking of returning from here when the depth reaches 0 (meaning that the outer `<extensions>` was closed. I don't know when the `for` loop finishes. Can there be something else after the final `</extensions>`?

If you look closely, the old code does a `return Ok(());`. Your first version did not.

ndrsg replied:
> If the XML library doesn't allow a mismatched closing tag, I think we shouldn't handle it here. Then the stack isn't necessary.

For code-clarity this should be handled here I guess, it should not cost much, hence the will not be huge extensions-tags. But if you want I can remove it of course.

> Instead I was thinking of returning from here when the depth reaches 0 (meaning that the outer <extensions> was closed. I don't know when the for loop finishes. Can there be something else after the final </extensions>?

I think the approach with counting the depth was not the best, hence it handles internal "extensions"-tags in a special way, different from the other internal tags.
I guess, the code should handle inner extensions-tags like any other inner tag, hence they depend on the extensions-schema, not the tag-name. The new code is much more generic and aligned with the coding-style of the other tag-parsers.
Something after the final </extensions> is (as a consequence of xml-schema and the visitor-pattern) always outside of the extensions-parser. This would have been the same with the depth approach.
This is then handled by the "outer" parser, in this case the track-parser and so on..

> If you look closely, the old code does a return Ok(());. Your first version did not.

Thats true, I got your point with returning ```Ok(())``` where you suggested it in the old code and I agree, this would have been correct.

lnicola replied:
Okay, so I tried to add a mismatched closing tag to a test, and got:

```
[src/parser/extensions.rs:77] &result = Err(
    XmlParseError(
        Error {
            pos: 5:22,
            kind: Syntax(
                "Unexpected closing tag: foo, expected extensions",
            ),
        },
    ),
)
```

And the `if let Some(current_inner_tag) = inner_tag_stack.pop()` check never sees the unmatched `</foo>. So `xml-rs` already handles this case, and I don't think we should do it a second time.

I also skimmed the rest of the code, and didn't find another place which builds a tag stack like this.

So if I'm not mistaken, that's dead code, and I think we should drop it. It's not just about the performance, but it's also misleading for the readers of the code.

My suggestion was to do something like:

```rust
/// consume consumes a single string as tag content.
pub fn consume<R: Read>(context: &mut Context<R>) -> GpxResult<()> {
    verify_starting_tag(context, "extensions")?;

    let mut depth = 1;
    for event in &mut context.reader {
        match event? {
            XmlEvent::StartElement { name, .. } => {
                // keep track of how many `extension` tags we are under
                if name.local_name == "extensions" {
                    depth += 1;
                }
            }
            XmlEvent::EndElement { name } => {
                dbg!(&name);
                if name.local_name == "extensions" {
                    // pop one
                    depth -= 1;
                }
                if depth == 0 {
                    return Ok(());
                }
            }
            _ => {}
        }
    }

    Err(GpxError::MissingClosingTag("extensions"))
}
```

Is this incorrect in some cases?

ndrsg replied:
> So if I'm not mistaken, that's dead code, and I think we should drop it. It's not just about the performance, but it's also misleading for the readers of the code.

Yes ok, I will remove it.


> Is this incorrect in some cases? Or is there any case when my version and yours produce different results?:

No, I do not think it is incorrect in an algorithmic way. It would work fine I guess.

I was just thinking, that all tags inside the extensions-tag should be treated the same way by the code. In the future the code should parse all the inner tags into a generic structure, that represents the all the tags inside the "extensions" to have it accessible for the user.
So in the future - if we migh have some generic-parser - we can, instead of 
```rust
            XmlEvent::StartElement { name, .. } => {
                // push every opening-element on the stack
                // we treat inner "extensions"-tags as any other tag
                let child = name.local_name.clone();
                inner_tag_stack.push(child);
            },
```

just write
```rust
            XmlEvent::StartElement { name, .. } => {
                // handle attributes etc ... 
                // ..
                let child = name.local_name.clone();
                generic_whatever::consume(child)
            },
```

and drop the whole inner stack/depth-tracking completely.


```rust
/// consume consumes a single string as tag content.
pub fn consume<R: Read>(context: &mut Context<R>) -> GpxResult<()> {
    verify_starting_tag(context, "extensions")?;

    let mut depth = 1;
    for event in &mut context.reader {
        match event? {
            XmlEvent::StartElement { name, .. } => {
                // I think its bad to hardcode the check on name == "extensions", because it is not a generic approach
                // and treats inner tags that are called "extensions" differently from any other inner tags, like "a", "foo", "bar"
                // It is correct, but feels wrong, maybe only a personal feeling
                if name.local_name == "extensions" {
                    depth += 1;
                }
            }
            XmlEvent::EndElement { name } => {
                if name.local_name == "extensions" {
                    // pop one
                    depth -= 1;
                    if depth == 0 {
                        return Ok(());
                    }
                }
            }
            _ => {}
        }
    }

    Err(GpxError::MissingClosingTag("extensions"))
}
```

I see the inner_tag vector more as a placeholder for an inner parser call. It gives a direction how I personally intended to go on with handling extensions inner content in another pullrequest based on this one.

In my opinion, depth tracking points in another direction, where the extensions-parser has to "know" something about the extensions content.

lnicola replied:
I'm not sure how that's supposed to work. This `consume` starts with an `<extensions>` and needs to parse (well, drop) everything until the matching `</extensions>`. We need to track the depth to achieve that (the parser won't do it for  us), regardless of the extension schema.

If we want to expose the contents of those to the user, agreed, that's not enough, but the stack is not helping with that, and it doesn't buy us _anything_, since the check you've added already exists in the parser.

lnicola replied:
I mean, the "ideal" API would be something like:

```rust
trait ExtensionParser {
    fn consume<R: Read>(&mut self, context: &mut Context<R>) -> GpxResult<()>;
}

struct Extensions;

impl ExtensionParser for Extensions {
    // ...
}

    let extension_parser = Extensions;
    let gpx: Gpx = GpxReader::new().with_extension_parser(&mut extension_parser).read(reader).unwrap();
```

, right?

And we don't need that stack to implement it.
